### PR TITLE
[13_3_X] Removal of StringCutObjectSelector from Muon trigger DQM

### PR DESCRIPTION
**Pull request by caruta**
#### PR description:

This PR removes the "StringCutObjectSelector" from the Muon trigger DQM, with the aim of solving the [memory issues](https://github.com/cms-sw/cmssw/issues/40437#issuecomment-1640459028) faced during the 2022 rereco. 
The cuts on the reco muons and trigger objects are now applied explicitly, since they are few and [identical](https://github.com/cms-sw/cmssw/issues/40437#issuecomment-1640560259) for all trigger paths.
  - Most of the changes are in [HLTMuonMatchAndPlot.cc](https://github.com/cms-sw/cmssw/compare/master...caruta:cmssw:master?expand=1#diff-def47c64859a25da5f880bada02a8b7995b19b6b65af9937e7a458e705304e3c) and in [HLTMuonMatchAndPlot.h](https://github.com/cms-sw/cmssw/compare/master...caruta:cmssw:master?expand=1#diff-342e3f99c9ba8c51d285b842040b9ea69152eb41a6a54ba073e91913d81c0429) to implement the new way of applying cuts on the muon and trigger objects
  - The configuration files [HLTMuonOfflineAnalyzer_cfi.py](https://github.com/cms-sw/cmssw/compare/master...caruta:cmssw:master?expand=1#diff-d8cb0507af5988917415877d046acc815ea6adc722532de0336faf9a13257331) and [HLTMuonOfflineAnalyzer_cff.py](https://github.com/cms-sw/cmssw/compare/master...caruta:cmssw:master?expand=1#diff-26e3f2f353a63e47667c31813450ca2ce5b485f5d61c2d314ae97c1dd9dc724e) have also been modified accordingly.

This PR will be backported down to the 124X release (used for the 2022 rereco).

#### PR validation:

Tested running the DQM module on some of the [latest RelVal](https://cmsweb.cern.ch/das/request?instance=prod/global&input=file+dataset%3D%2FRelValZMM_14%2FCMSSW_13_2_0_pre2-131X_mcRun3_2023_realistic_v6_2023-v1%2FGEN-SIM-RECO), and checked the agreement of the distributions. Also ran `runTheMatrix.py -l limited -i all --ibeos` and no failures were seen.

### CODE REVIEW

**Comment by emanueleusai** [comment]
Please test

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc
Comment (Dr15Jones) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:392]: This comment is no longer true

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc
Comment (Dr15Jones) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:404]: If set to true you can remove the else branch

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc
Comment (Dr15Jones) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:406]: Just do mu.isGlobalMuon()

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc
Comment (Dr15Jones) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:409]: You can just do mu.eta()

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc
Comment (Dr15Jones) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:436]: Can just do foundObject.eta()

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

**Comment by Dr15Jones** [comment]
Please remove commented out code.

**Comment by JanFSchulte** [comment]
@cms-sw/dqm-l2 Can the tests be run again and the review finished? This is an important fix to reduce memory consumption in ReReco jobs and quite urgent. Thanks!

**Comment by emanueleusai** [comment]
please test

**Comment by emanueleusai** [comment]
+1

- Differences in Muon HLT DQM plots as expected from the content of the PR
- Approving before last round of tests finishes due to urgency
- changes since last round of tests are only cosmetics (remove comments), so no changes in the results are expected

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc
Comment (perrotta) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:383]: These are all class members: I don't see the reason why to pass them as parameters to the method (it does not seem to me that they get modified by the method)
```suggestion
                                                  ) {
```

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc
Comment (perrotta) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:407]: These are all class members: I don't see the reason why to pass them as parameters to the method (it does not seem to me that they get modified by the method)
```suggestion
                                                                    ) {
```

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc
Comment (perrotta) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:395]: This could be made more compact:
```suggestion
    // minimun ID (requested for cosmics) is being a StandAlone muon
    bool muID = IsMuonGlb? mu.isGlobalMuon() : mu.isStandAloneMuon(); 
```

**Review by perrotta — COMMENTED** [review]
Please squash all commits into one when you are done with the updates

**Comment by perrotta** [comment]
> * Differences in Muon HLT DQM plots as expected from the content of the PR

Why there should be there differences in the HLT DQM plots? Shouldn't the same cuts being replicated identically as before here? (If not, please update the PR description to explicitate the modifications in the cut values)
@caruta @emanueleusai

**Comment by JanFSchulte** [comment]
@perrotta I don't see any differences in the HLT DQM in the latest set of tests. There seem to have been some in the first set of tests, but I think that this was a fluke because either baseline or reference (I can't ever tell from these plots which is which) was completely empty in one workflow. So I think we are actually no changes, as expected.

@slava77 I don't think we have explicitly checked and have just removed what we were pointed to as the offending aspect of the code. Can this be profiled as part of the PR tests, or do we need to do that offline?

**Comment by caruta** [comment]
Hi @slava77 (@JanFSchulte), I've tried to verify the memory consumption by dumping the stats results for all ME when running this DQM sequence on 100evts. Here are the results:
* Before the fix: 
    * Approx. maximum total virtual memory size of job: 2.63e+03 MB (reached 18 sec. after constructor called),
    * Avg virtual memory size of job: 2.57e+03 MB
* After the fix: 
    * Approx. maximum total virtual memory size of job: 2.44e+03 MB (reached 8 sec. after constructor called),
    * Avg virtual memory size of job: 2.34e+03 MB

I'm not sure this is enough and/or conclusive, but please let me know if you have other suggestions to test this!

**Comment by caruta** [comment]
> Please squash all commits into one when you are done with the updates

**Comment by perrotta** [comment]
> > Please squash all commits into one when you are done with the updates
> 
> @perrotta let me know if you're fine with the last commits, and if yes I'll proceed to propagate them to the other PRs for the back-porting. Thanks!

Thank you @caruta 
The latest uodates are fine with me
It would be nice if you could further squash the 15 commits into a single one

**Comment by JanFSchulte** [comment]
I guess then we should just enable profiling for the PR tests?

**Comment by slava77** [comment]
> I guess then we should just enable profiling for the PR tests?

I'd leave it to @perrotta  to decide; my past experience with the `enable profiling` is that they may slow down the test cycle to 24hrs and I'm not sure if the memory tests are working now.

**Comment by perrotta** [comment]
> > I guess then we should just enable profiling for the PR tests?
> 
> I'd leave it to @perrotta to decide; my past experience with the `enable profiling` is that they may slow down the test cycle to 24hrs and I'm not sure if the memory tests are working now.

We are in the week-end, and we can afford spending 24h for these tests, if we want to check it. Let do so.
In general, it can be much faster to run igprof privately on a relevant workflow and report here the result.
In any case, before merging the 15 commits must be squeezed into one.

**Comment by perrotta** [comment]
enable profiling

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
> **Additional Tests**: PROFILING

looking at [igprofMEM_GC_step3 in 136.889](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_13_3_X_2023-07-29-1100/el8_amd64_gcc11/profiling/136.889/PR-fce6b2/33970/igprofMEM_GC_step3.99) apparently the memory profiling does not work:
there are only 180 MB in MEM_LIVE.
I vaguely recall that this is an issue that appeared in EL8; and I guess there is no fix.

**Comment by caruta** [comment]
Sorry, my bad: trying to make the squash working I've messed up things... 
I can open a new PR with one clean commit with all the changes. Let me know if you agree! (@perrotta)

**Comment by perrotta** [comment]
> Sorry, my bad: trying to make the squash working I've messed up things... I can open a new PR with one clean commit with all the changes. Let me know if you agree! (@perrotta)

Perfect, thank you @caruta (and sorry for the extra work, but clean commits will definitely help future developers of the package

**Comment by caruta** [comment]
> > Sorry, my bad: trying to make the squash working I've messed up things... I can open a new PR with one clean commit with all the changes. Let me know if you agree! (@perrotta)
> 
> Perfect, thank you @caruta (and sorry for the extra work, but clean commits will definitely help future developers of the package). Please go ahead

A new PR has been created: [42437](https://github.com/cms-sw/cmssw/pull/42437).

**Comment by emanueleusai** [comment]
+1

**Pull request closed without merging**
